networkx/utils/decorators.py

```diff
@@ -450,22 +361,536 @@ def random_array(dims, random_state=1):
+        def foo(a, b, c):
+            return a, b, c
+
+    is equivalent to
+
+        def foo(a, b, c):
+            a, b = swap(a, b)
+            return a, b, z
+
+    Also, transforming functions can preceed a try-finally block, if they both
+    transform an argument and return a closing function:
+
+        def open_file(fn):
+            f = open(fn)
+            return f, lambda: f.close()
+
+        @argmap.try_finally(open_file, 'file')
+        def foo(file):
+            print(file.read())
+
+    is equivalent to
+
+        def foo(file):
+            file, close_file = open_file(file)
+            try:
+                print(file.read())
+            finally:
+                close_file()
+
+    """
+
+    def __init__(self, func, *args):
+        self._func = func
+        self._args = args
+        self._finally = False
+
+    @classmethod
+    def try_finally(cls, func, *args):
+        """Alternative decorator-constructor which first transforms one or more
+        arguments with func, executes the function to be decorated in a try
+        block, and then calls a cleanup function in the associated finally
+        block.  The cleanup function takes no arguments, and is expected to be
+        the second value returned by func.
+
+        Parameters
+        ----------
+        func : callable
+            The function to apply to arguments
+
+        *params: iterable of (int, str or tuple)
+            A list of parameters, specified either as strings (their names),
+            ints (numerical indices) or arbitrarily nested tuples thereof.
+
+        Examples
+        --------
+
+            def open_close(filename):
+                file = open(filename)
+                return file, lambda: file.close()
+
+            @argmap.try_finally(open_close, 0)
+            def print_file(file):
+                print(file.read())
+
+        generates a decorated function
+
+            def print_file(filename):
+                file, close = open_close(filename)
+                try:
+                    print(file.read())
+                finally:
+                    close()
+
+        """
+        dec = cls(func, *args)
+        dec._finally = True
+        return dec
+
+    @staticmethod
+    def _lazy_compile(func):
+        """Assemble and compile the source of our optimized decorator, and
+        intrusively replace its code with the compiled version's.
+
+        Parameters
+        ----------
+        func : callable
+            A function returned by argmap.__call__ which is in the process
+            of being called for the first time.
+
+        Returns
+        -------
+        func : callable
+            The same function, with a new __code__ object.
+
+        """
+        real_func = func.__argmap__.compile(func.__wrapped__)
+        func.__code__ = real_func.__code__
+        func.__globals__.update(real_func.__globals__)
+        func.__dict__.update(real_func.__dict__)
+        return func
```

Comment:
I have to admit I have very little experience with using/modifying these components of the object model directly.

Does the compilation have a significant impact on performance? Do we have a sense of what the impact of `_lazy_compile` is on a per-call basis vs. call number? My thinking here is if it's small (say O(10's of microseconds), arbitrarily) then maybe it's not worth the extra complexity.

Reply:
The issue highlighted in #4732 is that calling `compile` upon decoration significantly increases the import time.  The lazy strategy moves that cost to the first call of the decorated function.  Monkeypatching like this is not strictly necessary, and only serves to reduce the call-stack depth.  I'd be willing to back off on that if folks are uncomfortable with this.

Reply:
@dschult a second opinion, please?

Reply:
Ah - right I knew that conversation was somewhere - thanks for the link to the issue. All I was really looking for was the analysis that's already in that discussion :)

I'm happy with it as-is, I had been worried about other what happens when accessing/modifying these components of the object model for different Python implementations, but the PyPy tests in CI don't seem to mind so that significantly alleviates my concerns.

Reply:
I like the current implementation. Monkeypatching is always a little strange, but I think in this case the "lazy" really pays off for most people because it is rare that you use most of the functions in networkx and this cuts out all those compiles for the functions a user doesn't use. The first time compile is "paying the piper" but not with any significant interest in that only the code to set up the monkeypatch and do the compiling is the extra cost of delaying compiling.  After that it shouldn't make any difference (unless the monkeypatch call is slower than the automated lookup which I haven't checked, but I think is not slower).
:)